src/gt4py/next/ffront/ast_passes/single_static_assign.py

```diff
@@ -82,11 +179,78 @@ def apply(cls, node: ast.AST) -> ast.AST:
 
     def __init__(self, separator="__"):
         super().__init__()
-        self.name_counter: dict[str, int] = {}
-        self.separator: str = separator
+        self.versioning: Versioning = Versioning()
+        self.name_encoder: NameEncoder = NameEncoder(separator)
+
+    def _rename(self, node: ast.AST):
+        return self.RhsRenamer.apply(self.versioning, self.name_encoder, node)
+
+    def visit_FunctionDef(self, node: ast.FunctionDef):
+        # For practical purposes, this is sufficient, but really not general at all.
+        # However, the algorithm was never intended to be general.
+
+        old_versioning = self.versioning.copy()
 
-    def _rename(self, node):
-        return self.RhsRenamer.apply(self.name_counter, self.separator, node)
+        for arg in node.args.args:
+            self.versioning.define(arg.arg)
+
+        node.body = [self.visit(stmt) for stmt in node.body]
+
+        self.versioning = old_versioning
+        return node
+
+    def visit_If(self, node: ast.If) -> ast.If:
+        old_versioning = self.versioning
```

Comment:
If understand this correctly, the following `self._rename()` call cannot change the versioning dict, so why is this copy needed here?

Reply:
This is true, but you need `old_versioning` again for the _else_-branch after `self.versioning` has been changed for the _if_-branch.